crypto/evp/kdf_lib.c

```diff
@@ -144,6 +145,148 @@ int EVP_KDF_derive(EVP_KDF_CTX *ctx, unsigned char *key, size_t keylen,
     return ctx->meth->derive(ctx->algctx, key, keylen, params);
 }
 
+struct convert_key {
+    const char *name;
+    OSSL_PARAM *param;
+};
+
+static int convert_key_cb(const OSSL_PARAM params[], void *arg)
+{
+    struct convert_key *ckey = arg;
+    const OSSL_PARAM *raw_bytes;
+    unsigned char *data;
+    size_t len;
+
+    raw_bytes = OSSL_PARAM_locate_const(params, OSSL_SKEY_PARAM_RAW_BYTES);
+    if (raw_bytes == NULL)
+        return 0;
+
+    if (!OSSL_PARAM_get_octet_string_ptr(raw_bytes, (const void **)&data, &len))
+        return 0;
+
+    *ckey->param = OSSL_PARAM_construct_octet_string(ckey->name, data, len);
+    return 1;
+}
+
+int EVP_KDF_CTX_set_SKEY(EVP_KDF_CTX *ctx, EVP_SKEY *key, const char *paramname)
+{
+    struct convert_key ckey;
+
+    if (ctx == NULL)
+        return 0;
+
+    ckey.name = (paramname != NULL) ? paramname : OSSL_KDF_PARAM_KEY;
+
+    if (ctx->meth->set_skey != NULL) {
+        if (ctx->meth->prov != key->skeymgmt->prov) {
+            /* TODO: export/import dance */
+            return 0;
+        }
+        return ctx->meth->set_skey(ctx->algctx, key->keydata, ckey.name);
+    } else {
+        /*
+         * Provider does not support opaque keys, try to export and
+         * set params.
+         */
+        OSSL_PARAM params[2] = {
+            OSSL_PARAM_END,
+            OSSL_PARAM_END,
+        };
+        ckey.param = &params[0];
+
+        if (!ctx->meth->set_ctx_params)
+            return 0;
+
+        if (EVP_SKEY_export(key, OSSL_SKEYMGMT_SELECT_SECRET_KEY,
```

Comment:
something feels off about this.  The params array is set to OSSL_PARAMS_END for every entry, but convert_key_cb does this:
```
raw_bytes = OSSL_PARAM_locate_const(params, OSSL_SKEY_PARAM_RAW_BYTES);
```
and fails if the entry isn't found.  This seems like it will always fail.